Add capability to specify state args as a dict

bdrx312 opened the pull request:
Also add some more type hints and change some format calls to f strings. Fix state retry tests with invalid assumptions of duration and comment.

### What does this PR do?
Adds the ability to specify arguments to a state as a dictionary instead of having to specify a list of individual key: value pairs

### What issues does this PR fix or reference?
Fixes #68367

### Previous Behavior
Could only specify arguments to a state like this

```yaml
id1:
  cmd.run:
    - name: echo testing
    - umask: '0'
 ```

### New Behavior
Can specify arguments to a state like this

```yaml
id1:
  cmd.run:
    name: echo testing
    umask: '0'
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

twangboy reviewed (changes requested):
There are several locations in docs where we describe the state format. Are we going to recommend the new dict layout as opposed to the list of key/value pairs format? We should probably describe this new option in the docs as well. Here are a few locations:

https://github.com/saltstack/salt-user-guide/blob/main/docs/topics/states.rst

https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/starting_states.rst
https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt1.rst
https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt2.rst
https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt3.rst
https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt4.rst
https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt5.rst

Sxderp commented:
This is a pretty /fundamental/ change to Salt and probably a big uplift to keep around and supported. Before merging I think it would be prudent to dig through history and try and figure out why the list format.

bdrx312 commented on salt/state.py:
```diff
@@ -97,8 +101,6 @@
         "retry",
         "order",
         "parallel",
-        "prereq",
```
It is already in the `STATE_REQUISITE_KEYWORDS` set line 82.

bdrx312 commented:
> There are several locations in docs where we describe the state format. Are we going to recommend the new dict layout as opposed to the list of key/value pairs format? We should probably describe this new option in the docs as well. Here are a few locations:
> 
> https://github.com/saltstack/salt-user-guide/blob/main/docs/topics/states.rst
> 
> https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/starting_states.rst https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt1.rst https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt2.rst https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt3.rst https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt4.rst https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/states_pt5.rst

I was thinking this could go in as a kind trial first to make sure everything worked as desired before advertising it and then update the documentation to expose it. 

> Are we going to recommend the new dict layout as opposed to the list of key/value pairs format?

My personal thought/opinion is that the long term goal would to recommend the dict layout as opposed to the key/value pairs.

bdrx312 commented on tests/pytests/unit/state/test_state_compiler.py:
```diff
@@ -817,6 +814,34 @@ def test_render_requisite_require_in_disabled(minion_opts):
 
 
+def test_render_with_dict_args(minion_opts):
+    """Test disabling the require requisite via the options works"""
```
Fix test description comment.

lkubb commented on tests/pytests/pkg/integration/test_salt_call.py:
```diff
@@ -46,13 +46,36 @@ def state_name(salt_master):
     assert salt_master.is_running()
 
-    ret = salt_call_cli.run("state.apply", state_name)
+    ret = salt_call_cli.run("state.apply", fixture_name)
```
Sorry, I missed that the created state files are not called like the fixtures that create them. I'd recommend renaming the created sls files, but as a demonstration, this should work too:

```suggestion
@pytest.mark.parametrize("fixture_name", ["state_name", "state_name_dict_arg"])
def test_sls(salt_call_cli, salt_master, fixture_name, request):
    """
    Test calling a sls file
    """
    assert salt_master.is_running()
    sls_id = request.getfixturevalue(fixture_name)

    ret = salt_call_cli.run("state.apply", sls_id)
```

twangboy reviewed (changes requested):
I still think we should update the docs to show the new dict method in addition to the list method.

lkubb commented on salt/state.py:
```diff
@@ -72,9 +72,13 @@
+K = TypeVar("K")
+V = TypeVar("V")
+Pair = tuple[K, V]
+HighDataStateArgsDef = Union[Mapping[K, V], Iterable[Union[Mapping[K, V], str]]]
```
```suggestion
HighDataStateArgsDef = Mapping[K, V] | Iterable[Mapping[K, V] | str]
```

Minor suggestion: AFAIK 3008.0 is Python 3.10+ only, so you can use the `|` type operator. Unsure why `pyupgrade` does not catch this.

lkubb commented on salt/state.py:
```diff
@@ -440,75 +475,76 @@ def _verify_high(high: dict) -> list[str]:
+            )
+    elif arg_name in STATE_REQUISITE_KEYWORDS:
+        # if the arg is require or
+        # watch, it must be a list.
```
```suggestion
        # if the arg is a requisite,
        # it must be a list.
```

nit: I don't think it's specific to `require` and `watch`.

lkubb reviewed (commented):
Thanks for creating this! Just some minor comments.

bdrx312 commented on salt/state.py:
```diff
@@ -547,33 +583,27 @@ def pad_funcs(self, high):
-            skeys = set()
-            for key in sorted(high[name]):
+            state_keys = set()
+            for key, value in sorted(body.items(), key=lambda kv: kv[0]):
```
The lambda probably isn't necessary. I wanted to avoid the case where one of the other values in the tuple were not comparable. I am not sure off the top of my head what would happen in that case.

bdrx312 committed:
Add capability to specify state args as a dict

Also add some more type hints and change some format calls to f strings.
Fix state retry tests with invalid assumptions of duration and comment.

bdrx312 committed:
Replace Union type hint in state.py with |

bdrx312 committed:
Update documentation

bdrx312 commented on doc/ref/states/highstate.rst:
```diff
@@ -44,117 +44,66 @@ Example:
-Module reference
-----------------
+SLS module reference
+-------------
```
Oops, missed that after renaming. I started to name it `SLS reference`. Fixed.

bdrx312 commented:
> I still think we should update the docs to show the new dict method in addition to the list method.

@twangboy I updated the docs. Can you look over them and provide any feed back? I also have this question: https://github.com/saltstack/salt/pull/68368#issuecomment-3638372765

bdrx312 commented:
> What I meant by "add some docs" was to give some overall guidance on the different options for formatting the state, not going in to each state and changing the docs. https://docs.saltproject.io/salt/user-guide/en/latest/topics/states.html

I did update to give the different options https://github.com/saltstack/salt/pull/68368/files#diff-a0d563ea5979d3e412a93128f06d255fd9175f5ad9df06d193d6bcee9088af07R172-R201

I did not notice that page you linked; I will update that too. I also updated the examples because I think we want examples to show the recommended/ideal way to implement something. I find tools more difficult to use and frustrating when the tools improve over time and have better ways to configure and implement things but don't update the examples to use the newer methods.

As a unrelated side note but related to documentation, I started trying to on-board some team members so we can try to use salt and they found that user guide you linked, started at the beginning, and clicked through next on each page. When they got to the fourth page (State system behind the scenes https://docs.saltproject.io/salt/user-guide/en/latest/topics/state-system.html) they got a little overwhelmed by the information on that page and also didn't understand why they were already reading about what seemed like low level or implementation details. I think it would be good if that _State system behind the scenes_ came much later in the user guide once the basics were covered.

bdrx312 commented:
> I would say revert the test and add a new to so that both scenarios are tested.

There are other tests now in tests/pytests/functional/modules/state/test_state.py that are testing the new syntax. I don't think that one specifically needs to test both.

whytewolf commented:
> > I would say revert the test and add a new to so that both scenarios are tested.
> 
> There are other tests now in tests/pytests/functional/modules/state/test_state.py that are testing the new syntax. I don't think that one specifically needs to test both.

then revert the test. don't remove a test that was working fine for the old format.

whytewolf commented on doc/ref/states/extend.rst:
```diff
@@ -34,11 +80,40 @@ this:
+      ssh-server:
+        service:
+          - watch:
+            - file: /etc/ssh/banner
```
same as above. these document changes are not a part of the format change.

bdrx312 reviewed (commented):
> I will say one test i would like to see. is a single test that takes all three possible formats and checks the lowstate output comes out the same.

Good idea. Added

bdrx312 commented on doc/ref/states/highstate.rst:
```diff
@@ -346,58 +389,56 @@ components.
-
-
-    # inline function and names
-
```
```
# traditional declaration

    <ID Declaration>:
      <State Module Declaration>:
        - <Function Declaration>
        - <Function Arg Declaration>s...
        - <Name or Names Declaration>
        - <Requisite Declaration>s...
      <State Module Declaration>:
        - <Function Declaration>
        - <Function Arg Declaration>s...
        - <Name or Names Declaration>
        - <Requisite Declaration>s...
```

bdrx312 committed:
Add additional test for testing state formats and update more documentation to use new dict arguments format in examples

Revert change to an existing test back to using the old arguments format
Relax raising exception on invalid argument type in the state arguments to instead just log a warning and ignore the item

dwoz commented:
@whytewolf Are you good with this one now?

whytewolf commented:
with the coding of it I am good. I'm still not happy about the fundamental change itself. but that is not my call to make.
